File: R/Lesson_091_single_threaded_cpu.R
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 091 -- Single Threaded CPU
# Category   : Heap Priority Queue
# Difficulty : Medium
# Study Plan : Day 46
# =============================================================
#
# QUESTION:
#   You have tasks[i] = [enqueueTime, processingTime]. CPU picks the task with shortest processing time available; ties broken by index. Return order of indices.
# =============================================================

getOrder <- function(tasks){
  n <- length(tasks); idx <- order(sapply(tasks, function(x) x[1]))
  h <- list(); t <- 0; i <- 1; res <- c()
  while (i <= n || length(h) > 0){
    if (length(h)==0) t <- max(t, tasks[[idx[i]]][1])
    while (i <= n && tasks[[idx[i]]][1] <= t){
      h[[length(h)+1]] <- c(tasks[[idx[i]]][2], idx[i]); i <- i+1
    }
    pv <- sapply(h, function(x) x[1]); pi <- sapply(h, function(x) x[2])
    ord <- order(pv, pi); j <- ord[1]
    t <- t + h[[j]][1]; res <- c(res, h[[j]][2]); h[[j]] <- NULL
  }
  res
}
print(getOrder(list(c(1,2),c(2,4),c(3,2),c(4,1))))

File: Java/Lesson091_SingleThreadedCpu.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 091 -- Single Threaded CPU
// Category   : Heap Priority Queue
// Difficulty : Medium
// Study Plan : Day 46
// =============================================================
//
// QUESTION:
//   You have tasks[i] = [enqueueTime, processingTime]. CPU picks the task with shortest processing time available; ties broken by index. Return order of indices.
// =============================================================

import java.util.*;
public class Lesson091_SingleThreadedCpu {
    public int[] getOrder(int[][] tasks){
        Integer[] idx = new Integer[tasks.length];
        for (int i=0;i<idx.length;i++) idx[i]=i;
        Arrays.sort(idx,(a,b)->tasks[a][0]-tasks[b][0]);
        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b)->a[0]!=b[0]?a[0]-b[0]:a[1]-b[1]);
        long t=0; int i=0; int[] res = new int[tasks.length]; int k=0;
        while (i<idx.length || !pq.isEmpty()){
            if (pq.isEmpty()) t = Math.max(t, tasks[idx[i]][0]);
            while (i<idx.length && tasks[idx[i]][0]<=t){ pq.add(new int[]{tasks[idx[i]][1], idx[i]}); i++; }
            int[] x = pq.poll(); t += x[0]; res[k++] = x[1];
        }
        return res;
    }
    public static void main(String[] a){
        System.out.println(Arrays.toString(new Lesson091_SingleThreadedCpu().getOrder(new int[][]{{1,2},{2,4},{3,2},{4,1}})));
    }
}
